python/ray/air/checkpoint.py

```diff
@@ -359,7 +409,20 @@ def from_directory(cls, path: str) -> "Checkpoint":
         Returns:
             Checkpoint: checkpoint object.
         """
-        return cls(local_path=path)
+        state = {}
+
+        checkpoint_metadata_path = os.path.join(path, _CHECKPOINT_METADATA_FILE_NAME)
+        if os.path.exists(checkpoint_metadata_path):
+            with open(checkpoint_metadata_path, "rb") as file:
+                metadata = pickle.load(file)
+                cls._check_type_compatability(metadata.checkpoint_type)
+                cls = metadata.checkpoint_type
+                state = metadata.checkpoint_state
+
+        checkpoint = cls(local_path=path)
```

Comment:
We should be using `serialized_cls` here instead of `cls` right?
 
`serialized_cls` is the subclass based on this line: https://github.com/ray-project/ray/pull/28387/files#diff-5e1cc7107d4ba60e36c31b0b0bfa0bf8042c08b5dcad03dd8f21b68b5c19f2beR449.

So we should be restoring as the subclass, not as the superclass.

Basically this needs to work right?
```
ckpt = TorchCheckpoint.from_model(...)
assert type(Checkpoint.from_dict(ckpt.to_dict())) == TorchCheckpoint
```

Reply:
Can we also add a test for this?

Reply:
Oh, I see what you mean now. Yeah, this case won't work

```
ckpt = Checkpoint.from_dict(...)
assert type(TorchCheckpoint.from_dict(ckpt.to_dict())) == TorchCheckpoint
```

Reply:
Got it ok it's just 2 lines of common code, so ok to leave as is!